nncf/torch/sparsity/movement/structured_mask_handler.py

```diff
@@ -0,0 +1,211 @@
+        self.sparsifier_operand.weight_ctx.binary_mask = structured_mask_inflated
+        if self.sparsifier_operand.prune_bias is True:
+            self.sparsifier_operand.bias_ctx.binary_mask = structured_mask_inflated.amax(dim=1)
+
+
+class StructuredMaskHandler:
+
+    def __init__(self,
+                 sparsified_module_info_groups: List[SparsifiedModuleInfoGroup],
+                 strategy):
+        self.sparsified_module_info_groups = sparsified_module_info_groups
+        self.strategy = strategy
+        self.strategy_by_group_type = strategy.strategy_by_group_type
+        self._structured_mask_ctx_by_group_type = self._create_structured_mask_ctx_by_group_type()
+
+    def _get_group_of_prunable_sparsified_module_info(self) -> List[SparsifiedModuleInfoGroup]:
+        module_2_sparse_module_info_map = {sparse_info.module: sparse_info for sparse_info in self.sparsified_module_info}
+        building_blocks, _ = get_building_blocks(self.model,
+                                                 target_block_types=[BuildingBlockType.MSHA, BuildingBlockType.FF],
+                                                 block_filter_strategy=BlockFilteringStrategy.KEEP_SMALL,
+                                                 hw_fused_ops=True)
+        prunable_sparsified_module_info_groups = []
+        for group_id, building_block in enumerate(building_blocks):
+            sparsified_module_info = []
+            for op_addr in building_block.op_addresses:
+                if op_addr.operator_name in NNCF_MODULES_OP_NAMES:
+                    module = self.model.get_module_by_scope(op_addr.scope_in_model)
+                    module_info = module_2_sparse_module_info_map[module]
+                    sparsified_module_info.append(module_info)
+            prunable_sparsified_module_info_groups.append(
+                SparsifiedModuleInfoGroup(group_id,
+                                          building_block.block_type,
+                                          sparsified_module_info))
+        return prunable_sparsified_module_info_groups
+
+    def _create_structured_mask_ctx_by_group_type(self) -> List[Tuple[BuildingBlockType, List[StructuredMaskContext]]]:
+        structured_mask_ctx_by_group_type = []
+        for group in self.sparsified_module_info_groups:
+            group_type = group.group_type
+            ctxes = []
+            for module_info in group.sparse_module_info:
+                for rule in self.strategy_by_group_type[group_type]:
+                    if contains_any(module_info.module_node_name, rule.keywords):
+                        ctx = StructuredMaskContext(module_info.operand,
+                                                    module_info.module_node_name,
+                                                    rule.prune_grid)
+                        ctxes.append(ctx)
+                        break
+                else:
```

Comment:
@yujiepan-work this `else` seems out of place. Could you verify the intent? does the else correspond to any if?

Reply:
@yujiepan-work  It works fine, it was a bug at my end. Let's keep this, perhaps we can change the raised error message for clarity